### ISSUE TITLE
[8.0] Synchronize terms does "duplicate" user content translations

### ISSUE THREAD

**Issue by wtaferner** [odoo/odoo#9480]
## Impacted versions:
8.0 and above

## Steps to reproduce:
* **Use Case:** Single Language Shop Configuration (other language than English)
* Create several products and add content A on the field website_description in English (source)
* Switch language and translate the same field in any other language individually
* Synchronize terms

## Current behavior:
* The field translation w/ type model does something really weird...even if there is already a translation on an object the synchronize term does create a second translation (module is suddenly ```__export__```) based on the source term which actually does auto-retrieve a translation from a similar source term, but on the already existing translation.
* Based on this "duplication" the content gets polluted and the newly created content is wrongly auto-translated displayed.
* If the source term is hold in sync with target term it will just duplicate but the content is luckily also correct (workaround), based on the fact that the source is probably unique.

## Expected behavior:
* "Synchronize terms" does not pollute a translation which is not in sync with its co-incidentally similar source terms.
* If there is an existing translation on a type "model" field it will never create a second translation which will be then wrongly displayed based on the query and something like the first result

EDIT: Another and more detailed description you can find here https://github.com/odoo/odoo/issues/9480#issuecomment-161019522

OPW-654186

**Comment by wtaferner** [comment]
@mart-e 
In master we kind of fixed this for the workaround, but in v8 a customer of mine actually suddenly faced this issue and I needed to explain how to work it to circumvent this situation in the future.

Anyway, I think if we fix this in "Synchronize terms" such disasters won't happen to a whole website content translation (another language than English)

I leave it up to you as you know the implementation best and are responsible for translation stuff. Let me know if I can help...I need to give my customer the unpleasant hand to fix it on the instance itself.

**Comment by lapagept** [comment]
Same problem here when I synchronize terms. My french translations are duplicated. Not funny.

**Comment by lapagept** [comment]
For each translations, if I put myself the « module » value to « __export__ » at the creation of the translations, these translations are not duplicated when I « synchronize terms ». Hourra!

But I don't really understand why...

**Comment by wtaferner** [comment]
@lapagept 
It is clear if you copy and anticipate the behavior of the algorithm and assign the pseudo module ```__export__``` at least once that it does not duplicate which is normally used as module name for external_id assignment to know which objects have been exported to not duplicate the exported object on import.

Not sure what the strategy was/is but I guess it is about the translation for exported objects and the synchronize terms does replicate in case it was not done before. Pretty bad for a stable version though, but nobody is perfect and the bug is reported and I hope a solution and/or an official explanation is provided.

I guess if two entries for a single object and language are avoided/prohibited or the ```__export__``` module is not considered as real module there would not be an issue. @mart-e can you confirm what I have said? Still had no time to dive into the code and provide eventually a fix myself.

**Comment by mart-e** [comment]
Hi,

Sorry but I followed the given steps and I did not get duplication of translation.
Which revision of the code do you use?
Can you reproduce this any translatable field?

**Comment by wtaferner** [comment]
Hi Martin,
I saw your efforts and commits lately and I will try to give you a more clear reproduction (on runbot demo data) by looking deeper into and by retesting the issue based on your fixes on my instance first.

I am using as always the latest code and the issue mainly but not only came up for the website_description field of the product as described.

**Comment by mart-e** [comment]
Ok so if it's only website_description, it is probably due to the way the translations are created. Can you give the exact steps (which menu, actions) you used to create these ?

**Comment by lapagept** [comment]
It is a real problem here with one of our site.

**Comment by mart-e** [comment]
@lapagept then please tell us how to reproduce the issue on a standard odoo instance. Try on [runbot](http://runbot.odoo.com/runbot)

**Comment by wtaferner** [comment]
@mart-e 
Well, I had now the possibility to play around and I think I found the source of the issue and can name it.

There is a lot of duplication going on based on the column module of ```ir.translation```

## Steps to reproduce:

* Load a new language

## Use Case 1:

* Duplicate demo data (```product.product_product_6_product_template``` which you have translated before and you will get the result underneath for the newly created object

![9480-use-case-1](https://cloud.githubusercontent.com/assets/362478/11506200/7cedf3b0-984e-11e5-9231-97451a25a2a6.jpg)

## Use Case 2:

* Now export the new duplicated ```product.template``` and you will get the result underneath

![9480-use-case-2](https://cloud.githubusercontent.com/assets/362478/11506227/9ff086c0-984e-11e5-9682-a9478facddbe.jpg)

* Based on this delete all translations except the one without module defined (which means created purely by the user) and remember the product id to find it again ;-)

* Go to synchronize terms and wait...after synchronization you will get the extra translation with the module ```__export__``` back again b/c it is reference in ```ir.model.data``` after successful export of user-created data.

So, quite a lot to do I would say ;-)

Let me know if I might help further...

**Comment by mart-e** [comment]
Thanks for the detail, I have fixed it at 27d439901cd4f342a55d7c6ebf99b9b01770880d

This one was easy but I think we should stop worrying at these duplicated translations. While it's good to have only a single translation, this is the internal of Odoo. The issue would be if you get unexpected behaviour like getting the wrong translation for a record, but if, in the `ir.translation`, you a few translations linked to orphan records or duplicated, never mind.

**Comment by wtaferner** [comment]
Great that it was easy...thank you for the fix...I will test it although did you say that it is OK to have duplicate translations introduced by the system? Well, for this I have to say, it would not be acceptable and my customer had great issues (wrong translation was taken into account after synchronize terms)  and I had to fix it semi-automatically...but well, I will test it later and ask you to reopen if I do not agree with the improved behavior, ok?

**Comment by mart-e** [comment]
Getting wrong translations displayed to the customer would be an issue, I agree. Getting two identical translations in the ir.translation is not.

**Comment by wtaferner** [comment]
Ok, sure, two identical translations could be ignored although it is pretty bad for the customer if he/she wants to work on translation as I showed in the screenshot (clicking the translation button) b/c the user does not know which one will be the correct translation ^^

**Comment by lapagept** [comment]
Thanks guys

**Comment by wtaferner** [comment]
@mart-e 
Just did some tests...it is still inconsistent...

The ```product_template,website_description``` translation is ordered and displayed different compared to the ```product_template,name``` field, so there are two ways to fix this imho
* make sure it is working all the same way for all occurrences (no matter what) or 
* just get ridd of these useless and irritating duplicates.

Thank you for reopening.

**Comment by mart-e** [comment]
@wtaferner but how do you reproduce the issue with website_description? In which context do you get a different behaviour ?

**Comment by wtaferner** [comment]
Hi Martin,
I did now another test on runbot and the issue I still see for getting duplicates is the "Synchronize terms" algorithm. The use case is related to having an object successfully translated ones...

So, in case I have a user-generated content w/o module I will get another one with ```__export__``` afterwards, plus the wrong content is taken for duplication

## Use Case 1
* I have two user-generated objects (ids: 328,329)
* I export the object (id: 328)
* I duplicate the object (id: 328) which has id 329 (all good, the source term is still the same for both)
* Now I go to synchronize terms for the target language...and here you see the result...

* ID 328

![9480-use-case-5](https://cloud.githubusercontent.com/assets/362478/11556800/5cfcb5c6-99a9-11e5-84fe-671c1530acb6.jpg)

* ID 329

![9480-use-case-4](https://cloud.githubusercontent.com/assets/362478/11556794/58c97cc8-99a9-11e5-91b4-eaa692873ead.jpg)

So, firstly we have a duplication with the module ```__export__``` and secondly we have a wrong translation based on an unrelated object (ID 329) b/c of the same source term

## Use Case 2
Based on this there is the other issue where obviously on the website the order of translation taken to read and write is different than fields in the backend which can eventually be displayed in the frontend as well mentioned above. Test it yourself...having two translation for each field and the same object...change the content of the one with the higher ID and you will get two different results...one will be displayed unchanged and the other one will display the changed content.

I am pretty sure that is not a big deal and I could have provided a PR but based on the fact how much I had to explain I am happy that you do the fixes accordingly ;-)

**Comment by wtaferner** [comment]
@mart-e 
Could you reopen or do you prefer to have it transferred into a new issue?

**Comment by mart-e** [comment]
@wtaferner but in the first scenarios you give me, you don't show translation issues, you show that you get multiple `ir.translation` for the same object but not that you get a translation issue. If you translate (by switching the language of a user in his preference) an object with two `ir.translation`, you will see both will be updated.

For the second, I am not sure to understand because I can not reproduce your issue. I did the following:
1. create a product published on the ecommerce with a translation
2. get a second translation for this product (e.g. using the export + sync you show above)
3. make sure both translations are different
4. check the translation on backend and frontend => both are the same (we display the one with the higher ID)
Have I missed something?

**Comment by wtaferner** [comment]
@mart-e 
I think we should do a hangout...I am desperate...did you do exactly what I described?

Yes, I now know that if the user is doing a change of i.e. the name in the form view both translations are updated, but the issue is that the "Synchronize term" does create wrong translations (in the form of duplicates for once exported objects) and they are in effect as written on the website translation immediately after synchronize terms.

Obviously the most difficult thing to reproduce or to explain :-)
You have missed that "Synchronize terms" destroys a whole translation!! ;-)

**Comment by wtaferner** [comment]
> 4. check the translation on backend and frontend => both are the same (we display the one with the higher ID)
This is not true for the website_description...

**Comment by wtaferner** [comment]
@mart-e 
Please tell me that you got the issue...

**Comment by mart-e** [comment]
Ok I think I get the issue now, I needed to export the duplicated record too. I need a bit more investigation but reopening in the meantime.
Thanks

**Comment by wtaferner** [comment]
![giphy](https://cloud.githubusercontent.com/assets/362478/11590825/455c0a08-9a93-11e5-8c0e-fe76b184b11e.gif)

not the crude matter...

**Comment by wtaferner** [comment]
@mart-e 
I am very sorry, but the issue is still not fixed for the same use case I described several times.
Please reopen and let me test before you close it again. If you need further help to reproduce let us do a hangout as I think otherwise we will play this game like forever...

**Comment by mart-e** [comment]
I didn't close it, actually it auto-closed when @KangOl made the forward port to 9.0, due to "Fixes  #9480" in the commit message.
So yes the issue is still present, reopening.

**Comment by wtaferner** [comment]
Ok, thank you, I figured out later, tricky and lovely Github. I got auto-pranked ;-)

**Comment by wtaferner** [comment]
@mart-e 
As I just saw https://github.com/odoo/odoo/commit/6a366b216b3d119cd5546c877a2f77b5342cb452 by @beledouxdenis which is probably needed most likely b/c of this issue I was asking myself what happened to the progress? Any news?

OPW was closed with the argument that it is not wanted to maintain two points/ways of communication, but as long it is fixed in a reasonable time I said I will not complain :-) It is a critical issue as described therefore I am kind of getting impatient now ;-)

**Comment by flotho** [comment]
Hi @wtaferner an @mart-e 

I have exactly the same issue in v9. How do you work to solve this?
Do you delete the newest translations which are duplicates.

regards

**Comment by rco-odoo** [comment]
Fixed in 8.0 by a46f8bb9c502c9e82fb1392aeb6120eaeb8d98bf.

**Comment by bealdav** [comment]
cleaning: 
```sql
DELETE FROM ir_translation
WHERE id IN (
    SELECT id
    FROM (SELECT id,
        ROW_NUMBER() OVER (partition
            BY module, name, res_id, lang, state, type, comments, src, value
            ORDER BY id) AS rnum
        FROM ir_translation) t
        WHERE t.rnum > 1);

```

Source: https://wiki.postgresql.org/wiki/Deleting_duplicates
